osu.Game/Skinning/SkinProvidingContainer.cs

```diff
@@ -39,26 +43,47 @@ public class SkinProvidingContainer : Container, ISkinSource
 
         protected virtual bool AllowColourLookup => true;
 
+        /// <summary>
+        /// Constructs a new <see cref="SkinProvidingContainer"/> with a single skin added to the protected <see cref="SkinSources"/> list.
+        /// </summary>
         public SkinProvidingContainer(ISkin skin)
+            : this()
         {
-            this.skin = skin;
+            SkinSources.Add(skin);
+        }
 
+        /// <summary>
+        /// Constructs a new <see cref="SkinProvidingContainer"/> with no sources.
+        /// Up to the implementation for adding to the <see cref="SkinSources"/> list.
+        /// </summary>
+        protected SkinProvidingContainer()
+        {
             RelativeSizeAxes = Axes.Both;
 
             noFallbackLookupProxy = new NoFallbackProxy(this);
 
-            if (skin is ISkinSource source)
-                source.SourceChanged += TriggerSourceChanged;
+            SkinSources.BindCollectionChanged(((_, args) =>
+            {
+                switch (args.Action)
+                {
+                    case NotifyCollectionChangedAction.Add:
+                        foreach (var source in args.NewItems.Cast<ISkin>().OfType<ISkinSource>())
+                            source.SourceChanged += OnSourceChanged;
+
+                        break;
+
+                    case NotifyCollectionChangedAction.Remove:
+                        foreach (var source in args.OldItems.Cast<ISkin>().OfType<ISkinSource>())
+                            source.SourceChanged -= OnSourceChanged;
+
+                        break;
+                }
+            }), true);
         }
 
         public ISkin FindProvider(Func<ISkin, bool> lookupFunction)
         {
-            if (skin is ISkinSource source)
-            {
-                if (source.FindProvider(lookupFunction) is ISkin found)
-                    return found;
-            }
-            else if (skin != null)
+            foreach (var skin in SkinSources)
```

Comment:
This is very weird to me, because something like `c.FindProvider(s => s.GetTexture(...) != null)` is actually going to iterate through `SkinSources` twice. Once in here and once inside the `lookupFunction`.

It seems like this iteration may not be required?

Reply:
Also it's not actually consuming the `skin` here..

Reply:
Whoops... this actually needs some work, I was hoping i could get rid of `FindProvider` but looks like that's not happening anytime soon. (with animations and taiko mascot animations being the most blocking one)

Reply:
Resolved with a refactor of this disability logic, now the allow checks are performed on a private `DisableableSkinSource` and all lookups in `SkinProvidingContainer` are forwarded to them (which includes `FindProvider` which now does `lookupFunction(disableableSkinSource[skin])`).